src/lib.rs

```diff
@@ -565,8 +614,71 @@ pub fn get_lines_from_osm(pbf: &mut OsmPbfReader) -> Vec<Line> {
         .collect()
 }
 
+pub fn get_routes_from_stop(routes: &Vec<Route>, stop_point: &StopPoint) -> Vec<Route> {
+    routes
+        .iter()
+        .filter(|route| route.contains_stop_point_id(&stop_point.id))
+        .map(|obj| obj.clone())
+        .collect()
+}
+
+pub fn categorize_stop_point(stop_point: &StopPoint, routes: Vec<Route>) -> StopPoint {
+    let mut result_sp = stop_point.clone();
+    if result_sp
+        .all_osm_tags
+        .contains("public_transport", "platform")
+    {
+        result_sp.stop_point_type = StopPointType::Platform;
+    } else if result_sp
+        .all_osm_tags
+        .contains("public_transport", "stop_position")
+    {
+        result_sp.stop_point_type = StopPointType::StopPosition;
+    } else {
+        let mut routes_ptv2: Vec<Route> = routes
+            .iter()
+            .filter(|r| r.all_osm_tags.contains("public_transport:version", "2"))
+            .map(|obj| obj.clone())
+            .collect();
+        warn!(
+            "categorization of stop_point {} needs pt_v2 routes. {} ptv2 routes found",
+            stop_point.id,
+            routes_ptv2.len()
+        );
+        routes_ptv2.sort_by(|a, b| b.id.cmp(&a.id));
+        for route in routes_ptv2 {
+            let ptv2_stop_point_uses = route.get_stop_point_roles(&stop_point.id);
+            if ptv2_stop_point_uses.contains(&String::from("platform"))
+                || ptv2_stop_point_uses.contains(&String::from("platform_exit_only"))
+                || ptv2_stop_point_uses.contains(&String::from("platform_entry_only"))
+            {
+                result_sp.stop_point_type = StopPointType::Platform;
+            } else if ptv2_stop_point_uses.contains(&String::from("stop"))
+                || ptv2_stop_point_uses.contains(&String::from("stop_exit_only"))
+                || ptv2_stop_point_uses.contains(&String::from("stop_entry_only"))
+            {
+                result_sp.stop_point_type = StopPointType::StopPosition;
+            }
+            if result_sp.stop_point_type != StopPointType::Unknown {
+                break;
+            }
+        }
+    }
+    result_sp
+}
+
+pub fn update_stop_points_type(
+    stop_points: &Vec<StopPoint>,
+    routes: &Vec<Route>,
+) -> Vec<StopPoint> {
+    stop_points
+        .iter()
+        .map(|sp| categorize_stop_point(sp, get_routes_from_stop(&routes, sp)))
+        .collect()
+}
```

Comment:
Instead of returning the same list of `StopPoint` (by reallocating the memory space), you can take a `&mut` reference on `stop_points` (some other part of the code will need to change too). See 
```suggestion
pub fn update_stop_points_type(
    stop_points: &mut Vec<StopPoint>,
    routes: &Vec<Route>,
) {
    stop_points.iter_mut().for_each(|mut sp| {                   
        let route_from_stops = get_routes_from_stop(&routes, sp);
        categorize_stop_point(&mut sp, route_from_stops)         
    })                                                           
}
```
See https://github.com/prhod/osm-transit-extractor/pull/1.